Sources/Moya/Task.swift

```diff
@@ -3,39 +3,48 @@ import Foundation
 /// Represents an HTTP task.
 public enum Task {
 
-    /// A request with no additional data.
-    case requestPlain
-
-    /// A requests body set with data.
-    case requestData(Data)
-
-    /// A request body set with `Encodable` type
-    case requestJSONEncodable(Encodable)
-
-    /// A request body set with `Encodable` type and custom encoder
-    case requestCustomJSONEncodable(Encodable, encoder: JSONEncoder)
-
-    /// A requests body set with encoded parameters.
-    case requestParameters(parameters: [String: Any], encoding: ParameterEncoding)
-
-    /// A requests body set with data, combined with url parameters.
-    case requestCompositeData(bodyData: Data, urlParameters: [String: Any])
-
-    /// A requests body set with encoded parameters combined with url parameters.
-    case requestCompositeParameters(bodyParameters: [String: Any], bodyEncoding: ParameterEncoding, urlParameters: [String: Any])
-
-    /// A file upload task.
-    case uploadFile(URL)
-
-    /// A "multipart/form-data" upload task.
-    case uploadMultipart([MultipartFormData])
-
-    /// A "multipart/form-data" upload task  combined with url parameters.
-    case uploadCompositeMultipart([MultipartFormData], urlParameters: [String: Any])
-
-    /// A file download task to a destination.
-    case downloadDestination(DownloadDestination)
-
-    /// A file download task to a destination with extra parameters using the given encoding.
-    case downloadParameters(parameters: [String: Any], encoding: ParameterEncoding, destination: DownloadDestination)
+    /// All different ways to set parameters in an HTTP request's body.
+    public enum BodyParams {
+        /// The given encodable will be set in the request's body without any additional encoding.
+        case raw(Data)
+
+        /// The given encodable will be json encoded in the request's body.
+        case json(Encodable, JSONEncoder = JSONEncoder())
+
+        /// The given encodable will be url encoded in the request's body.
+        case urlEncoded(Encodable, URLEncodedFormEncoder = URLEncodedFormEncoder())
+
+        /// The given encodable will be encoded in the request's body using the provided encoder.
+        ///
+        /// The provided encoder must not be a `URLEncodedFormParameterEncoder` (use `BodyParams.urlEncoded`instead)
+        /// or `JSONParameterEncoder`(use `BodyParams.json` instead). If this is the case, a `MoyaError.encodableMapping` will be raised.
+        case custom(Encodable, ParameterEncoder)
+    }
+
+    /// All different ways to set parameters in an HTTP request's query.
+    public enum QueryParams {
+        /// The given encodable will be url encoded in the request's query.
+        case query(Encodable, URLEncodedFormEncoder = URLEncodedFormEncoder())
```

Comment:
```suggestion
        case urlEncoded(Encodable, URLEncodedFormEncoder = URLEncodedFormEncoder())
        case custom(Encodable, ParameterEncoder)
```

Reply:
Do you have a use case for the .custom value? I struggle to find one way to add parameters to an url without it being with the query. Or is it just for the purpose of providing a fully customizable way, "just in case of" ?

Reply:
yeah I just try to leave a room for things we didn't think about at the moment but users will probably find a use case to use their own parameter encoder

Reply:
I've just commited some changes regarding custom params.
The issue with your proposition is that it would result in a `custom` value in both `BodyParams` and `Queryparams`, while they would work exactly the same.
So I've created a `CustomParams` struct we can now provide as associated value of the `Task` value and removed the `BodyParams.custom` value.
At the same time, `QueryParams` is now a struct, which is less weird than an enum with only one case.